### PR TITLE
[feature] Ability to query API resources that match any or multiple words of search term

### PR DESCRIPTION
**Pull request by alis-khadka**
Addresses: https://github.com/restarone/violet_rails/issues/1374

**DEMO**

https://user-images.githubusercontent.com/25191509/214291392-23e3c962-31ce-4df3-b3fa-1a2565c978be.mov

### CODE REVIEW

File: test/controllers/api/resource_controller_test.rb [hunk 1]
```diff
@@ -285,6 +285,25 @@ class Api::ResourceControllerTest < ActionDispatch::IntegrationTest
...
+  test '#index search jsonb field - string - partial: multi word string' do
+    @api_resource_1.update(properties: {name: 'Professional Writer'})
+    @api_resource_2.update(properties: {name: 'Physical Development'})
+    @api_resource_3.update(properties: {name: 'Professional Development'})
+
+    payload = { 
+      properties: { 
+        name: { 
+          value: 'professional development',
+          option: 'PARTIAL'
```
Comment (donrestarone) [inline test/controllers/api/resource_controller_test.rb:297]: our use case for `PARTIAL` is something like this `%professional development%`. so adding a new matcher for `KEYWORDS` would be better where, 

1. if a string is passed ("Pro development") => "pro" and "development"
2. if an array of strings are passed `['pro', 'dev']` => "pro" and "dev"

Comment (alis-khadka) [reply test/controllers/api/resource_controller_test.rb:297]: > if an array of strings are passed ['pro', 'dev'] => "pro" and "dev"
@donrestarone For confirmation;

For ['pro dev', 'test'], both 'pro' & 'dev' should match, right?

Comment (donrestarone) [reply test/controllers/api/resource_controller_test.rb:297]: yup! @alis-khadka

**Comment by donrestarone** [comment]
Code LGTM @alis-khadka -- 

@Ayon95 could you please update the car app (https://violet-test.net/car-renderer) to allow filter logic customization? I have noted my suggestion below: 

 
Screen Shot 2023-01-24 at 2 15 35 PM

The goal is to test out Keyword search and validate the other search options we have

**Comment by Ayon95** [comment]
@donrestarone I've added the matcher dropdown and filter logic. Can you clarify how strict searching will work? There are three search options - PARTIAL, KEYWORDS, and EXACT.

https://violet-test.net/car-renderer

**Comment by donrestarone** [comment]
@Ayon95 strict should be strictly equal. In the case below it is not working as expected: 

Screen Shot 2023-01-25 at 4 01 02 PM

also, the keyword search is not working as expected. It should show results from GMA and Audi in the query below: 
Screen Shot 2023-01-25 at 4 00 22 PM

**Comment by Ayon95** [comment]
Have the changes in this PR been deployed to Violet testnet?

**Comment by donrestarone** [comment]
@Ayon95 no its not deployed to the permanent test app (https://violet-test.net) I was hoping for you to write the client code there, so I could get that code by simply relaunching the review app for this PR

**Comment by Ayon95** [comment]
@donrestarone The different types of search should work in the review app

**Comment by kripananda-yadav** [comment]
Acceptance Criteria: 

When user enters query for a search like : "Buy Cars 2023"

- Then Result must contains: Results for Words like "Buy", "Cars", and "2023" **(KEYWORDS)**
- Results must contain: Result for Words like "Buy Cars", "Buy 2023" and "Cars 2023" **( PARTIAL)**
- Results must contain: Result for Word like "Buy Cars 2023" **(EXACT)**

**Comment by kripananda-yadav** [comment]
Deployed Review app is not accessible 

Screenshot 2023-01-28 at 2 43 01 PM

**Comment by kripananda-yadav** [comment]
@alis-khadka  The review app page specifically (https://review-1379.violet-test.net/car-renderer) is unreachable now, It was reachable before.
Screenshot 2023-01-28 at 3 40 48 PM

**Comment by kripananda-yadav** [comment]
@alis-khadka  @donrestarone The above issue was debugged and I have raised a issue.
https://github.com/restarone/violet_rails/issues/1388

**Comment by kripananda-yadav** [comment]
@alis-khadka When user selects Matcher or Strict in the Search filter it is automatically changing to Partial.

https://user-images.githubusercontent.com/70195786/215271968-499165ff-b69e-4a88-a8ac-2fac825025e3.mov

**Comment by Ayon95** [comment]
@kripananda-yadav You have to specify the matcher before entering the search term in the search bar. If no matcher is selected before searching, partial searching will be the default behavior.

How partial searching in the backend works is that it tries to find API resources that have at least the specified search term. For example, if "Buy Cars 2023" is searched with "PARTIAL" option, then API resources that have values like "Buy Cars 2023", "Buy Cars 2023 xyz" will be matched.

The way strict or exact searching is different from partial searching is that it matches API resources that have exactly the specified search term (case insensitive), not more, not less.

**Comment by donrestarone** [comment]
when a blank string is passed to the keyword matcher, the following exception is raised: 

``` txt
An ActionView::Template::Error occurred in content#show:

  PG::SyntaxError: ERROR:  syntax error at or near "OR"
LINE 1: ...roperties ->> 'published') = lower('true')) AND ( OR  OR ) O...
                                                             ^

  (erb):1:in `parse_dynamic_attribute'
```

**Comment by alis-khadka** [comment]
> when a blank string is passed to the keyword matcher, the following exception is raised:
> 
> ```
> An ActionView::Template::Error occurred in content#show:
> 
>   PG::SyntaxError: ERROR:  syntax error at or near "OR"
> LINE 1: ...roperties ->> 'published') = lower('true')) AND ( OR  OR ) O...
>                                                              ^
> 
>   (erb):1:in `parse_dynamic_attribute'
> ```

I have pushed some changes to handle this on the backend side. I think it should also have been handled from the client side  when the values are blank strings.

**Comment by kripananda-yadav** [comment]
@donrestarone @alis-khadka https://github.com/restarone/violet_rails/issues/1194 Due to this issue the assets url are  wrong.
Screenshot 2023-02-14 at 3 05 25 PM

**Comment by alis-khadka** [comment]
@donrestarone 
Please have a look at this.
https://github.com/restarone/violet_rails/issues/1194#issuecomment-1454999788

**Pull request merged**